File: push-client-javascript/socket_channel.js
const net = require('net')
const socketFactory = require('./socket_factory')

class SocketChannel {

  constructor(socket,client){
    this.socket = socket
    this.dataQueue = []
    this.client = client

    this.socket.on('data',data => {
      const datas = data.toString().split('\n')
      for(let i = 0;i<datas.length;i++){
        if (datas[i]) {
          this.client.onMessage(datas[i])
        }
      }
    })
  }

  writeAndFlush(msg){
    this.socket.write(msg)
  }

  close(){
    this.socket.destroy()
  }
}

module.exports = { SocketChannel }

File: push-client-go/client/socket_channel.go
package client

type SocketChannel struct {
	socket  *Socket
}

func NewSocketChannel(socket *Socket) *SocketChannel {
	sc := new(SocketChannel)
	sc.socket = socket
	return sc
}

func (sc *SocketChannel) WriteAndFlush(msg string) error {
	return sc.socket.Write([]byte(msg))
}

func (sc *SocketChannel) ReadLine() (string, error) {
	line, err := sc.socket.ReadLine()
	return string(line), err
}